### PR TITLE
Logger refactoring

### PR DESCRIPTION
**Pull request by ckyang**

### CODE REVIEW

**Comment by ckyang** [comment]
Please find attached file for your detail
/home/ckyang/Presentation/LoggerRefactoring.odp

**Comment by ckyang** [comment]
[](url)

**Comment by ckyang** [comment]
Please find attached file for detail logger refactoring.
[LoggerRefactoring.pdf](https://github.com/Zilliqa/Zilliqa/files/1897240/LoggerRefactoring.pdf)

**Comment by jiayaoqijia** [comment]
Hi Clark, nice proposal. Can you take a look at the popular logging frameworks below ?Then we can choose the most suitable one for us. Thanks.  https://github.com/gabime/spdlog/blob/master/README.md and 
https://github.com/KjellKod/g3log
https://github.com/google/glog

**Comment by ckyang** [comment]
Hi Yaoqi,

OK, I will check and update soon

**Comment by ckyang** [comment]
https://github.com/KjellKod/g3log/issues/153

It seems there is limitation on g3log that it can has only one instance by design.
Need to investigate the solution on generating state-xxx-log.txt & zilliqa-xxx-log.txt.

**Comment by ckyang** [comment]
Updated document:
[LoggerRefactoring.pdf](https://github.com/Zilliqa/Zilliqa/files/1915021/LoggerRefactoring.pdf)

**Comment by ckyang** [comment]
Updated document
[LoggerRefactoring.pdf](https://github.com/Zilliqa/Zilliqa/files/1922356/LoggerRefactoring.pdf)

**Comment by ckyang** [comment]
@Gnnng Good suggestion, I will use cmake ExternalProject for g3log.

File: src/libUtils/Logger.h [hunk 1]
```diff
@@ -124,12 +141,46 @@ class ScopeMarker
...
         std::ostringstream oss;                                                \
         oss << msg;                                                            \
         Logger::GetStateLogger(NULL, true)                                     \
             .LogState(oss.str().c_str(), __FUNCTION__);                        \
     }
-
+#define LOG_GENERAL(level, msg)                                                \
+    {                                                                          \
+        std::ostringstream oss;                                                \
+        oss << msg;                                                            \
```
Comment (jiayaoqijia) [inline src/libUtils/Logger.h]: Do we need put std::ostringstream oss for every function?

**Pull request merged**